### PR TITLE
Reentrancy with call target

### LINKED ISSUE #3047: Reentrant projects with CallTarget can fail without error
This is a reduced repro from the internal VS repo.

### Steps to reproduce

a.proj

```xml
<Project>

<Target Name="EntryTarget">
  <MSBuild Projects="b.proj;c.proj" BuildInParallel="true" />
</Target>

</Project>
```

b.proj

```xml
<Project>

<Target Name="BTarget">
  <MSBuild Projects="Microsoft.VisualStudio.Shell.csproj" Targets="BuildGenerateSources" BuildInParallel="true" />
</Target>

</Project>
```

c.proj

```xml
<Project>

<Target Name="CTarget">
  <Exec Command="ping 127.0.0.1 -n 1" />
  <MSBuild Projects="Microsoft.VisualStudio.Shell.csproj" Targets="BuildGenerated" BuildInParallel="true" />
</Target>

</Project>
```

delay.proj

```xml
<Project>

<Target Name="Delay">
  <Exec Command="ping 127.0.0.1 -n 2" />
</Target>

</Project>
```

Microsoft.VisualStudio.Shell.csproj

```xml
<Project DefaultTargets="Build">
<Target Name="BuildGenerateSources" DependsOnTargets="_Get;Build">
</Target>

<Target Name="BuildGenerated">
  <CallTarget Targets="Build" />
</Target>


<Target Name="Build">
  <CallTarget Targets="_Get" />
</Target>

<Target Name="_Get">
  <MSBuild Projects="delay.proj" BuildInParallel="true" />
  <Exec Command="ping 127.0.0.1 -n 5" YieldDuringToolExecution="true" StandardOutputImportance="low" />
</Target>

</Project>
```

Command line
```
MSBuild.exe a.proj /m
```
### Expected  behavior

Builds with no errors.

### Actual behavior

```
Microsoft (R) Build Engine version 15.6.82.30579 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 3/4/2018 2:55:06 PM.
     1>Project "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\a.proj" on node 1 (default targets).
     1>Project "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\a.proj" (1) is building "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\b.proj" (2) on node 1 (default targets).
     2>Project "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\b.proj" (2) is building "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\Microsoft.VisualStudio.Shell.csproj" (3) on node 1 (BuildGenerateSources target(s)).
     3>Project "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\Microsoft.VisualStudio.Shell.csproj" (3) is building "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\delay.proj" (4) on node 1 (default targets).
     4>Delay:
         ping 127.0.0.1 -n 2
     1>Project "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\a.proj" (1) is building "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\c.proj" (5) on node 2 (default targets).
     5>CTarget:
         ping 127.0.0.1 -n 1
     4>Done Building Project "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\delay.proj" (default targets).
     3>_Get:
         ping 127.0.0.1 -n 5
     3>Done Building Project "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\Microsoft.VisualStudio.Shell.csproj" (BuildGenerateSources target(s)) -- FAILED.
     2>Done Building Project "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\b.proj" (default targets) -- FAILED.
     5>Done Building Project "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\c.proj" (default targets) -- FAILED.
     1>Done Building Project "S:\work\intermittent-devdiv-failure\super-simplified-repro\multi_active - Copy\a.proj" (default targets) -- FAILED.

Build FAILED.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:05.68
```

Note that the overall build and serveral targets failed, but no tasks failed and there are no logged errors.

### PR DESCRIPTION
**Pull request by cdmihai**
fixes #3047

### CODE REVIEW

**Comment by AndyGerlicher** [comment]
Hrm, `Microsoft.Build.UnitTests.EscapingInProjects_Tests.SimpleScenarios.SemicolonInPropertyPassedIntoITaskItemParam_UsingTaskHost` is the failure I was looking at in my other PR that I wasn't able to repro!

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for Full please

**Comment by cdmihai** [comment]
@dotnet-bot test this please

**Comment by cdmihai** [comment]
Don't merge yet. I am going to add a few more commits

File: src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs [hunk 1]
```diff
@@ -456,12 +456,19 @@ internal static IBuildComponent CreateComponent(BuildComponentType type)
...
+                        // the target, encountered a legacy CallTarget, pushed new work onto the stack, and yielded back to here. Instead of starting the already-partially-
+                        // built target, wait for the other one to complete. Then CheckSkipTarget will skip it here.
+                        bool wasActivelyBuilding = await CompleteOutstandingActiveRequests(currentTargetEntry.Name);
+
                         // It's possible that our target got pushed onto the stack for one build and had its dependencies process, then a re-entrant build came in and
                         // actually built this target while we were waiting, so that by the time we get here, it's already been finished.  In this case, just blow it away.
                         if (!CheckSkipTarget(ref stopProcessingStack, currentTargetEntry))
                         {
+                            //ErrorUtilities.VerifyThrow(!wasActivelyBuilding, "Target {0} was actively building and waited on but we are attempting to build it again.", currentTargetEntry.Name);
```
Comment (rainersigwald) [inline src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs]: Should we bring this back?

Comment (cdmihai) [reply src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs]: Yes, forgot about it :)

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 2]
```diff
@@ -1547,7 +1547,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
       than one framework.
    -->
    <ItemGroup>
-      <_MSBuildProjectReferenceExistent Condition="'%(_MSBuildProjectReferenceExistent.SkipGetTargetFrameworkProperties)' == '' and '%(Extension)' == '.vcxproj'">
+      <_MSBuildProjectReferenceExistent Condition="'%(_MSBuildProjectReferenceExistent.SkipGetTargetFrameworkProperties)' == '' and ('%(Extension)' == '.vcxproj' or '%(Extension)' == '.nativeproj')">
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets]: @AndyGerlicher you wanted to just pull this in, right?

**Comment by rainersigwald** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please

https://ci2.dot.net/job/Microsoft_msbuild/job/master/job/innerloop_Windows_NT_CoreCLR_prtest/222/

```
Microsoft.Build.UnitTests.GenerateResource_Tests.InProc.RequiredTransformations.ForceSomeOutOfDate
Microsoft.Build.UnitTests.GenerateResource_Tests.InProc.RequiredTransformations.ForceOutOfDate
```

**Pull request merged**